### PR TITLE
484 incorporate enumerations

### LINKED ISSUE #484: Enumeration incorporation
Co-authored-by: Benjamin Gorman <benjamin-gorman@uiowa.edu>

### Is your feature request related to a problem? Please describe.

Iteration of [#403] [#407]

- many parameters are constrained to a subset of str
- similarly named parameters have different subsets
- docstrings are not uniform in format and content

### Describe the solution you'd like

- create enumerations
  - in monai/utils/enums.py
  - use internally to ensure correctness
  - allows us to type subsets
  - allows us to enforce subsets
- rename parameters
  - name the parameters to match the underlying function call
  - allows for transparency
- standardize docstrings
  - use set notation to specify the str subset
  - include 'Defaults to' to specify default value
  - include 'See also' to specify underlying function
  - allows for clear communication of subset and function call
- type hint
  - accept both the str and enum
  - allows us to differentiate similarly named parameters

#### Docstring, Enumeration, Type Hint

- the docstring format could be changed (style choice)
- some enumeration class names should be changed (need suggestions)
- some enumerations may not be worth it

##### mode [1]

```python
"""
mode: {``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``, ``"mean"``,
    ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
    One of the listed string values or a user supplied function. Defaults to ``FIXME``.
    See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
"""

class NumpyPadMode(Enum):
    """
    See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
    """

    CONSTANT = "constant"
    EDGE = "edge"
    LINEAR_RAMP = "linear_ramp"
    MAXIMUM = "maximum"
    MEAN = "mean"
    MEDIAN = "median"
    MINIMUM = "minimum"
    REFLECT = "reflect"
    SYMMETRIC = "symmetric"
    WRAP = "wrap"
    EMPTY = "empty"


Union[NumpyPadMode, str]
Optional[Union[NumpyPadMode, str]]
Union[Sequence[object], NumpyPadMode, str]
```

##### mode [2]

```python
"""
mode: {``"bilinear"``, ``"nearest"``}
    Interpolation mode to calculate output values. Defaults to ``FIXME``.
    See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
"""

class GridSampleMode(Enum):
    """
    See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
    """

    BILINEAR = "bilinear"
    NEAREST = "nearest"


Union[GridSampleMode, str]
Optional[Union[GridSampleMode, str]]
Union[Sequence[object], GridSampleMode, str]
```

##### mode [3]

```python
"""
mode: {``"nearest"``, ``"linear"``, ``"bilinear"``, ``"bicubic"``, ``"trilinear"``, ``"area"``}
    The interpolation mode. Defaults to ``FIXME``.
    See also: https://pytorch.org/docs/stable/nn.functional.html#interpolate
"""

class InterpolateMode(Enum):
    """
    See also: https://pytorch.org/docs/stable/nn.functional.html#interpolate
    """

    NEAREST = "nearest"
    LINEAR = "linear"
    BILINEAR = "bilinear"
    BICUBIC = "bicubic"
    TRILINEAR = "trilinear"
    AREA = "area"


Union[InterpolateMode, str]
Optional[Union[InterpolateMode, str]]
Union[Sequence[object], InterpolateMode, str]
```

##### mode [4]

```python
"""
mode: {``"nearest"``, ``"linear"``, ``"bilinear"``, ``"bicubic"``, ``"trilinear"``}
    If ends with ``"linear"`` will use ``spatial dims`` to determine the correct interpolation.
    This corresponds to linear, bilinear, trilinear for 1D, 2D, and 3D respectively.
    The interpolation mode. Defaults to ``FIXME``.
    See also: https://pytorch.org/docs/stable/nn.html#upsample
"""

class UpsampleMode(Enum):
    """
    See also: https://pytorch.org/docs/stable/nn.html#upsample
    """

    NEAREST = "nearest"
    LINEAR = "linear"
    BILINEAR = "bilinear"
    BICUBIC = "bicubic"
    TRILINEAR = "trilinear"


Union[UpsampleMode, str]
Optional[Union[UpsampleMode, str]]
Union[Sequence[object], UpsampleMode, str]
```

##### mode [5]

```python
"""
mode: {``"constant"``, ``"gaussian"``}
    How to blend output of overlapping windows. Defaults to ``FIXME``.

    - ``"constant``": gives equal weight to all predictions.
    - ``"gaussian``": gives less weight to predictions on edges of windows.
"""

class BlendMode(Enum):
    """
    See also: :py:class:`monai.data.utils.compute_importance_map`
    """

    CONSTANT = "constant"
    GAUSSIAN = "gaussian"


Union[BlendMode, str]
Optional[Union[BlendMode, str]]
Union[Sequence[object], BlendMode, str]
```

##### padding_mode [6]

```python
# the underlying function call has it named 'mode'
# only called in 'sliding_window_inference' where it conflicts with 'mode [5]'
# naming it 'mode' for this function would be misleading anyway
"""
padding_mode: {``"constant"``, ``"reflect"``, ``"replicate"``, ``"circular"``}
    Padding mode when ``roi_size`` is larger than inputs. Defaults to ``FIXME``
    See also: https://pytorch.org/docs/stable/nn.functional.html#pad
"""

class PytorchPadMode(Enum):
    """
    See also: https://pytorch.org/docs/stable/nn.functional.html#pad
    """

    CONSTANT = "constant"
    REFLECT = "reflect"
    REPLICATE = "replicate"
    CIRCULAR = "circular"


Union[PytorchPadMode, str]
Optional[Union[PytorchPadMode, str]]
Union[Sequence[object], PytorchPadMode, str]
```

##### padding_mode [7]

```python
"""
padding_mode: {``"zeros"``, ``"border"``, ``"reflection"``}
    Padding mode for outside grid values. Defaults to ``FIXME``.
    See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
"""

class GridSamplePadMode(Enum):
    """
    See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
    """

    ZEROS = "zeros"
    BORDER = "border"
    REFLECTION = "reflection"


Union[GridSamplePadMode, str]
Optional[Union[GridSamplePadMode, str]]
Union[Sequence[object], GridSamplePadMode, str]
```

##### average [8]

```python
"""
average: {``"macro"``, ``"weighted"``, ``"micro"``, ``"none"``}
    Type of averaging performed if not binary classification. Defaults to ``FIXME``.

    - ``"macro"``: calculate metrics for each label, and find their unweighted mean.
        This does not take label imbalance into account.
    - ``"weighted"``: calculate metrics for each label, and find their average,
        weighted by support (the number of true instances for each label).
    - ``"micro"``: calculate metrics globally by considering each element of the label
        indicator matrix as a label.
    - ``"none"``: the scores for each class are returned.
"""

class Average(Enum):
    """
    See also: :py:class:`monai.metrics.rocauc.compute_roc_auc`
    """

    MACRO = "macro"
    WEIGHTED = "weighted"
    MICRO = "micro"
    NONE = "none"


Union[Average, str]
Optional[Union[Average, str]]
Union[Sequence[object], Average, str]
```

##### reduction [9]

```python
"""
reduction: {``"none"``, ``"mean"``, ``"sum"``, ``"mean_batch"``, ``"sum_batch"``, ``"mean_channel"``, ``"sum_channel"``}
    Define the mode to reduce computation result of 1 batch data. Defaults to ``FIXME``.
"""

class MetricReduction(Enum):
    """
    See also: :py:class:`monai.metrics.meandice.DiceMetric`
    """

    NONE = "none"
    MEAN = "mean"
    SUM = "sum"
    MEAN_BATCH = "mean_batch"
    SUM_BATCH = "sum_batch"
    MEAN_CHANNEL = "mean_channel"
    SUM_CHANNEL = "sum_channel"


Union[MetricReduction, str]
Optional[Union[MetricReduction, str]]
Union[Sequence[object], MetricReduction, str]
```

##### reduction [10]

```python
"""
reduction: {``"none"``, ``"mean"``, ``"sum"``}
    Specifies the reduction to apply to the output. Defaults to ``FIXME``.

    - ``"none"``: no reduction will be applied.
    - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
    - ``"sum"``: the output will be summed.
"""

class LossReduction(Enum):
    """
    See also:
        - :py:class:`monai.losses.dice.DiceLoss`
        - :py:class:`monai.losses.dice.GeneralizedDiceLoss`
        - :py:class:`monai.losses.focal_loss.FocalLoss`
        - :py:class:`monai.losses.tversky.TverskyLoss`
    """

    NONE = "none"
    MEAN = "mean"
    SUM = "sum"


Union[LossReduction, str]
Optional[Union[LossReduction, str]]
Union[Sequence[object], LossReduction, str]
```

##### w_type [11]

```python
"""
w_type: {``"square"``, ``"simple"``, ``"uniform"``}
    Type of function to transform ground truth volume to a weight factor. Defaults to ``FIXME``.
"""

class Weight(Enum):
    """
    See also: :py:class:`monai.losses.dice.GeneralizedDiceLoss`
    """

    SQUARE = "square"
    SIMPLE = "simple"
    UNIFORM = "uniform"


Union[Weight, str]
Optional[Union[Weight, str]]
Union[Sequence[object], Weight, str]
```

##### norm_type [12]

```python
"""
norm_type: {``"batch"``, ``"instance"``}
    Feature normalisation with batchnorm or instancenorm. Defaults to ``FIXME``.
"""

# other 'norm_type' are using 'monai.networks.layers.factories.Norm'
class Normalisation(Enum):
    """
    See also:
        - :py:class:`monai.networks.nets.ConvNormActi`
        - :py:class:`monai.networks.nets.HighResBlock`
        - :py:class:`monai.networks.nets.HighResNet`
    """

    BATCH = "batch"
    INSTANCE = "instance"


Union[Normalisation, str]
Optional[Union[Normalisation, str]]
Union[Sequence[object], Normalisation, str]
```

##### acti_type [13]

```python
"""
acti_type: {``"relu"``, ``"prelu"``, ``"relu6"``}
    Non-linear activation using ReLU or PReLU. Defaults to ``FIXME``.
"""

# other 'acti_type' are using 'monai.networks.layers.factories.Act'
class Activation(Enum):
    """
    See also:
        - :py:class:`monai.networks.nets.ConvNormActi`
        - :py:class:`monai.networks.nets.HighResBlock`
        - :py:class:`monai.networks.nets.HighResNet`
    """

    RELU = "relu"
    PRELU = "prelu"
    RELU6 = "relu6"


Union[Activation, str]
Optional[Union[Activation, str]]
Union[Sequence[object], Activation, str]
```

##### channel_matching [14]

```python
"""
channel_matching: {``"pad"``, ``"project"``}
    Specifies handling residual branch and conv branch channel mismatches. Defaults to ``FIXME``.

    - ``"pad"``: with zero padding.
    - ``"project"``: with a trainable conv with kernel size.
"""

class ChannelMatching(Enum):
    """
    See also: :py:class:`monai.networks.nets.HighResBlock`
    """

    PAD = "pad"
    PROJECT = "project"


Union[ChannelMatching, str]
Optional[Union[ChannelMatching, str]]
Union[Sequence[object], ChannelMatching, str]
```

##### method [15]

```python
"""
method: {``"symmetric"``, ``"end"``}
    Pad image symmetric on every side or only pad at the end sides. Defaults to ``FIXME``.
"""

class Method(Enum):
    """
    See also: :py:class:`monai.transforms.croppad.array.SpatialPad`
    """

    SYMMETRIC = "symmetric"
    END = "end"


Union[Method, str]
Optional[Union[Method, str]]
Union[Sequence[object], Method, str]
```

### Describe alternatives/additions you've considered

- deriving enumerations from both str and Enum
  - e.g. ``class Mode(str, Enum):``
  - loses the typing benefit of enumerations
- leaving subset differentiation to the docstring
- leaving parameter's names as is
- typing to not allow a mix of str and enum in a sequence
  - e.g. ``Union[Sequence[Mode], Sequence[str], Mode, str]``
- create and use type aliases

### PR DESCRIPTION
**Pull request by benjamin-gorman**
Fixes #484.

### Description

The proposed solution as described in #484.

- subsets are enforced by an enum's "constructor"
- enums are used internally

**Note**

- some enumeration class names should be changed (need suggestions)
- some enumerations may not be worth it

### Status

**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

**Comment by ericspod** [comment]
Is it feasible to change [LayerFactory](https://github.com/Project-MONAI/MONAI/blob/f7309a8e1d0a75d81aaa1e49917a106e73fca2d0/monai/networks/layers/factories.py#L70) to inherit Enum as well? It defines a map from names to layer types, if the names could be treated as Enum members the factories and their names can be used in the type hints.

**Comment by benjamin-gorman** [comment]
> Is it feasible to change [LayerFactory](https://github.com/Project-MONAI/MONAI/blob/f7309a8e1d0a75d81aaa1e49917a106e73fca2d0/monai/networks/layers/factories.py#L70) to inherit Enum as well? It defines a map from names to layer types, if the names could be treated as Enum members the factories and their names can be used in the type hints.

LayerFactory seems to be very dynamic; I'm not sure I understand how Enum would be incorporated. It would be a different approach to make an enumeration for each layer type.

File: monai/data/grid_dataset.py [hunk 1]
```diff
@@ -44,7 +48,7 @@ def __init__(self, dataset, patch_size, start_pos=(), mode: str = "wrap", **pad_
         self.dataset = dataset
         self.patch_size = (None,) + tuple(patch_size)
         self.start_pos = start_pos
-        self.mode = mode
+        self.mode = NumpyPadMode(mode)
```
Comment (hjmjohnson) [inline monai/data/grid_dataset.py:51]: `self.mode: NumpyPadMode = NumpyPadMode(mode)`
To explicitly type self.mode.  While the __init__ may be a string or a NumpyPadMode, internally only NumpyPadMode is vaild.

Comment (benjamin-gorman) [reply monai/data/grid_dataset.py:51]:
```python
self.mode = NumpyPadMode(mode)
reveal_type(self.mode)
```

```bash
$> mypy monai/data/grid_dataset.py 
monai/data/grid_dataset.py:52: note: Revealed type is 'monai.utils.enums.NumpyPadMode*'
```

Internally mypy already places that restriction (the asterisk denotes the type was inferred).

Being explicit would serve to protect the type by having mypy throw an error if the ``NumpyPadMode()`` were to be removed.

**Comment by ericspod** [comment]
My idea was to change LayerFactory to inherit Enum, the names would then show up as Enum values so we could type them as the factory they came from. I don't know if that'll work with the factories being instances rather than types themselves.

File: monai/data/grid_dataset.py (hunk 1 above)
Comment (hjmjohnson) [reply monai/data/grid_dataset.py:51]: @benjamin-gorman I am a code safety, explicit is better than implicit, junkie.  I would prefer that all "revealed types" were explicit.

**Comment by benjamin-gorman** [comment]
> My idea was to change LayerFactory to inherit Enum, the names would then show up as Enum values so we could type them as the factory they came from. I don't know if that'll work with the factories being instances rather than types themselves.

Yeah, the factories being instances will be a problem if we try to inherit Enum: the constructor turns into a lookup function. LayerFactory could inherit Enum and define all the necessary functions. Then a Layer would inherit LayerFactory and define the members. That would [follow the rules](https://docs.python.org/3/library/enum.html#restricted-enum-subclassing) and allow for typing. It would not allow for dynamic adding of members though.

**Comment by benjamin-gorman** [comment]
> Excellent changes overall. One comment, should be explicitly set the type hint for `self.XXXX: Enumtype = Enumtype(stringequv)`

Changed

**Pull request merged**